src/query/service/src/pipelines/processors/transforms/runtime_pool.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Mutex;
+
+pub trait RuntimeBuilder<T> {
+    type Error;
+    fn build(&self) -> Result<T, Self::Error>;
+}
+
+pub struct Pool<R, B: RuntimeBuilder<R>> {
+    pub builder: B,
+    runtimes: Mutex<Vec<R>>,
+}
+
+impl<R, B: RuntimeBuilder<R>> Pool<R, B> {
+    pub fn new(builder: B) -> Self {
+        Self {
+            builder,
+            runtimes: Mutex::new(vec![]),
```

Comment:
Why not generate some runtimes during initialisation, the default number of threads is the same as the number of cpus, will the initial runtime number 0 cause a scramble on the first run and affect performance?

Reply:
For udf, the maximum number of instances is the number of threads; for udaf it's a bit more complicated because udaf will have multiple phases.
The current runtime's lifecycle is within one sql. What is the benefit of early initialization? The benefit of delayed initialization is obvious: it is possible that a few instances are enough. This implementation is a copy of arrow_udf_wasm, which is why wasm doesn't have a pool, so it might make sense to put the runtime pool inside arrow_udf, for example, to save parse script ast every time.

Reply:
Ok, may be we can contribute this feature to `arrow_udf`.